### PR TITLE
feature(thumbnails): add the ability to define custom image processors

### PR DESCRIPTION
**Pull request by fschade**
## Description
Thumbnails can now be changed during creation, previously the images were always scaled to fit the given frame,
but it could happen that the images were cut off because they could not be placed better due to the aspect ratio.

This pr introduces the possibility of specifying how the behavior should be, following processors are available

* resize
* fit
* fill
* thumbnail

the processor can be applied by adding the processor query param to the request, e.g. `processor=fit`, `processor=fill`, ...

to find out more how the individual processors work please read https://github.com/disintegration/imaging

if no processor is provided it behaves the same as before (resize for gif's and thumbnail for all other)

needs web changes too, next step....

## Related Issue
- Fixes https://github.com/owncloud/enterprise/issues/6057
- Fixes https://github.com/owncloud/ocis/issues/5179
- Prep for https://github.com/owncloud/web/issues/7728
- Web counterpart https://github.com/owncloud/web/pull/9806

## Motivation and Context
web preview app is not just in use to preview images, but also to actually consume the content, therefor it is needed that images are un-cropped for that, in the file-list for example it is fine to have cropping in place.

## How Has This Been Tested?
- unit tests
- manual with postman

## Screenshots (if appropriate):

without processor:
Screenshot 2023-10-04 at 16 34 15

with fit processor:
Screenshot 2023-10-04 at 16 34 41

## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [X] Code changes
- [X] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

could need some docs magic cc.: @mmattel

### CODE REVIEW

**Comment by mmattel** [comment]
@fschade pls add your description to the readme of the thumbnail service, thx

File: services/thumbnails/README.md [hunk 1]
```diff
@@ -45,9 +45,20 @@ Various resolutions can be defined via `THUMBNAILS_RESOLUTIONS`. A requestor can
...
+Returned: 15x10
+
+## Thumbnail Processors
+
+Image generation can be configured by defining different processors, following processors are available:
+
+* resize
+* fit
+* fill
+* thumbnail
```
Comment (mmattel) [inline services/thumbnails/README.md:59]:
```suggestion
*   resize
*   fit
*   fill
*   thumbnail
```
Fixing codacy complains

File: services/thumbnails/pkg/thumbnail/generator.go [hunk 2]
```diff
@@ -11,36 +10,33 @@ import (
...
 	}
 
-	return imaging.Thumbnail(m, size.Dx(), size.Dy(), imaging.Lanczos), nil
+	return processor.Process(m, size.Dx(), size.Dy(), imaging.Lanczos), nil
 }
 
+// GifGenerator is used to create a web friendly version of the provided gif image.
 type GifGenerator struct{}
 
-func (g GifGenerator) GenerateThumbnail(size image.Rectangle, img interface{}) (interface{}, error) {
```
Comment (kobergj) [inline services/thumbnails/pkg/thumbnail/generator.go:38]: exported func `Generate` should have comment or be unexported

File: services/thumbnails/pkg/thumbnail/processor.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+// ID returns the processor identification.
+func (p DefinableProcessor) ID() string { return p.Slug }
+
+// Process transforms the given image.
+func (p DefinableProcessor) Process(img image.Image, width, height int, filter imaging.ResampleFilter) *image.NRGBA {
+	return p.Converter(img, width, height, filter)
+}
+
+// ProcessorFor returns a matching Processor
+func ProcessorFor(id, fileType string) (Processor, error) {
```
Comment (kobergj) [inline services/thumbnails/pkg/thumbnail/processor.go:31]: Imho this should return a `DefinableProcessor` (return structs, expect interfaces) But this is very nitpicky I admit

**Review by kobergj — CHANGES_REQUESTED** [review]

File: services/thumbnails/pkg/thumbnail/generator.go [hunk 4]
```diff
@@ -11,36 +10,33 @@ import (
...
+func (g GifGenerator) Generate(size image.Rectangle, img interface{}, processor Processor) (interface{}, error) {
```
Comment (kobergj) [inline services/thumbnails/pkg/thumbnail/generator.go:34]: exported func Generate should have comment or be unexported

**Review by kobergj — CHANGES_REQUESTED** [review]

**Comment by tbsbdr** [comment]
@kobergj If you have the chance, could you (or someone else) review this PR? (I'm really looking forward having this improvement/fix). thanks @fschade for the implementation!

File: services/thumbnails/README.md [hunk 5]
```diff
@@ -45,9 +45,20 @@ Various resolutions can be defined via `THUMBNAILS_RESOLUTIONS`. A requestor can
...
+## Thumbnail Processors
+
+Image generation can be configured by defining different processors, following processors are available:
+
+*   resize
+*   fit
+*   fill
+*   thumbnail
+
+to apply one of those, a query parameter has to be added to the request, e.g. `?processor=fit`
```
Comment (mmattel) [inline services/thumbnails/README.md:61]:
```suggestion
*   `resize`
*   `fit`
*   `fill`
*   `thumbnail`

To apply one of those, a query parameter has to be added to the request, e.g. `?processor=fit`
```
Just a minor finding I missed before

**Comment by mmattel** [comment]
LGTM from a docs pov 👍

File: services/thumbnails/README.md [hunk 6]
```diff
@@ -45,9 +45,20 @@ Various resolutions can be defined via `THUMBNAILS_RESOLUTIONS`. A requestor can
...
+## Thumbnail Processors
+
+Image generation can be configured by defining different processors, following processors are available:
+
+*   `resize`
+*   `fit`
+*   `fill`
+*   `thumbnail`
+
+To apply one of those, a query parameter has to be added to the request, e.g. `?processor=fit`
```
[the remark this answers is not in the document]
Comment (mmattel) [reply services/thumbnails/README.md:61]: If I recall correctly, this is optional, means behaviour equals before if not added.
Maybe to write: `Image generation can optionally be configured...`
But @fschade can clarify best.

**Comment by fschade** [comment]
@michaelstingl scalingup has no effect if i remember correctly, `a` is new to me, i can add a fallback if you want, but please let me add it in a separate pr.

File: services/thumbnails/README.md [hunk 7]
```diff
@@ -45,9 +45,20 @@ Various resolutions can be defined via `THUMBNAILS_RESOLUTIONS`. A requestor can
 
 Example:
 
-Requested: 18x12  
-Available: 30x20, 15x10, 9x6  
-Returned: 15x10  
+Requested: 18x12
+Available: 30x20, 15x10, 9x6
+Returned: 15x10
```
[the remark this answers is not in the document]
Comment (fschade) [reply services/thumbnails/README.md:50]: not sure, i have to have a look into the codebase, to be honest, i do not know every thumbnail service detail. please create a issue if we should change something.

File: services/thumbnails/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (fschade) [reply services/thumbnails/README.md:61]: if no processor is provided it uses the default behaviour as before (Resize fort gif, Thumbnail for everything else)

File: services/thumbnails/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (michaelstingl) [reply services/thumbnails/README.md:50]: Okay, I’ll play with it and I’ll open new issues if needed.

**Comment by michaelstingl** [comment]
Okay, I’ll play with it and I’ll open new issues if needed.

**Review by kobergj — APPROVED** [review]

**Pull request merged**